README.md
# Pypleskapi

Easy requests to [Parallels Plesk Panel API RPC](http://www.parallels.com/products/plesk/)

## About
Instead of building complex structures of XML, you can write all the requests in dict types.
Pypleskapi converts dict type packets into XML structures, and so as the other way around,
you just need to know how to write the corresponding dict structure.

    >>> from pleskapi import StructDict
    >>> from pleskapi import send_packet

    >>> packet = StructDict('1.6.3.5')
    >>> packet['webspace']['get']['filter'] = {'name' : 'domain.tld'}
    >>> packet['webspace']['get']['dataset']['gen_info']
	
	>>> print packet.dict()
	{'packet': {'webspace': {'get': {'filter': {'name': 'domain.tld'}, 'dataset': {'gen_info': {}}}}, '@version': '1.6.3.5'}}

	>>> print packet.xml(True)
	<?xml version="1.0" encoding="UTF-8"?>
	<packet version="1.6.3.5">
	  <webspace>
	    <get>
    	  <filter>
	        <name>domain.tld</name>
	      </filter>
	      <dataset>
	        <gen_info/>
	      </dataset>
	    </get>
	  </webspace>
	</packet>

    >>> r = send_packet(packet.dict())
    >>> r.response()
    {'version': '1.6.3.5', 'result': {'status': 'ok', 'filter-id': 'domain.tld', 'id': '36'}}

StructDict builds chain's of dict without expliciting, common dict's are build this way:
	
	>>> d = dict() or {}
	>>> d['webspace'] = {}
	{'webspace': {}}

You need to create a new dict for every key:

	>>> d['webspace'] = {}
	>>> d['webspace']['get'] = {}
	{'webspace': {'get': {}}}
	>>> d['webspace']['get']['filter'] = { 'name' : 'domain.tld' }
	Traceback (most recent call last):
	  File "<stdin>", line 1, in <module>
	KeyError: 'get'

With StructDict you can build chain's of dict:
	
	>>> from pleskapi import StructDict
	""" 1.6.3.5 refer to  packet header body - <packet version="1.6.3.5">...</packet> """
	>>> sd = StructDict('1.6.3.5')
	>>> sd['webspace']['get']['filter']
	{}
	>>> sd
	{'webspace': {'get': {'filter': {}}}}
	>>> print sd.xml()
	<?xml version="1.0" encoding="UTF-8"?><packet version="1.6.3.5"><webspace><get><filter/></get></webspace></packet>

Reference: [Plesk Api](http://download1.parallels.com/Plesk/PP11/11.0/Doc/en-US/online/plesk-api-rpc/33899.htm)
The conversion are based in [Hay's project](http://github.com/hay/xml2json)

## Installation

TODO

## Prerequisites

TODO

## Overview

Before starting this topic, I recommend the reading of [API RPC Manual](http://www.parallels.com/download/plesk/11/documentation/) for
better understanding how it works.
For building dict structures that will become valid requests to Plesk API RPC, you need to understand how a dict represents an XML structure.
The conversion follow the example bellow:

<pre>
XML                              JSON
&lt;e/&gt;                             "e": null
&lt;e&gt;text&lt;/e&gt;                      "e": "text"
&lt;e name="value" /&gt;               "e": { "@name": "value" }
&lt;e name="value"&gt;text&lt;/e&gt;         "e": { "@name": "value", "#text": "text" }
&lt;e&gt; &lt;a&gt;text&lt;/a &gt;&lt;b&gt;text&lt;/b&gt; &lt;/e&gt; "e": { "a": "text", "b": "text" }
&lt;e&gt; &lt;a&gt;text&lt;/a&gt; &lt;a&gt;text&lt;/a&gt; &lt;/e&gt; "e": { "a": ["text", "text"] }
&lt;e&gt; text &lt;a&gt;text&lt;/a&gt; &lt;/e&gt;        "e": { "#text": "text", "a": "text" }
</pre>

Reference: [Converting Between XML and JSON](http://www.xml.com/pub/a/2006/05/31/converting-between-xml-and-json.html)

An easy way of understanding it's using the converter functions, you can convert from an XML structure to a python dict type

	>>> from pleskapi import converter as conv
	>>> xmlstr = '<?xml version="1.0" encoding="UTF-8"?><packet version="1.6.3.5"><webspace><get><filter/></get></webspace></packet>'
	>>> conv.xml2dict(xmlstr)
	{'packet': {'webspace': {'get': {'filter': None}}, '@version': '1.6.3.5'}}

TODO: PleskApiError

Use help bult-in for more info about the methods and classes.

	>>> from pleskapi.base import BaseRequest, send_packet, BaseResponse
	>>> help(BaseRequest)
	>>> help(send_packet)
	>>> help(BaseResponse)

## Ordering Dict's

Plesk RPC API needs that the XML structure follow a specific order, more info: [API RPC Manual](http://www.parallels.com/download/plesk/11/documentation/) - API RPC > API RPC Packets > How to Create Packets.
A python dict type is unordered, so you need to use an OrderedDict type for ordering only the necessary keys.
Let's consider the XML string bellow:

	<?xml version="1.0" encoding="UTF-8"?>
	<packet version="1.6.3.5">
	  <ip>
	    <add>
	      <ip_address>192.0.2.18</ip_address>
	      <netmask>255.255.255.0</netmask>
	      <type>shared</type>
	      <interface>eth0</interface>
	    </add>
	  </ip>
	</packet>

ip_address node must be the first, netmask the second and so on.
Building this structure without using OrderedDict, outputs to:
	
	>>> from pleskapi import StructDict as sd
	
	>>> sd['ip']['add'] = {}
	>>> sd['ip']['add']['ip_address'] = '192.0.2.18'
	>>> sd['ip']['add']['netmask'] = '255.255.255.0'
	>>> sd['ip']['add']['type'] = 'shared'
	>>> sd['ip']['add']['interface'] = 'eth0'
	>>> sd.dict()
	{'ip': {'add': {'interface': 'eth0', 'type': 'shared', 'netmask': '255.255.255.0', 'ip_address': '192.0.2.18'}}}

The keys in 'add' are unordered, so the XML structure will be:

	<?xml version="1.0" encoding="UTF-8"?>
	<packet version="1.6.3.5">
	  <ip>
	    <add>
	      <interface>eth0</interface>
	      <type>shared</type>
	      <netmask>255.255.255.0</netmask>
	      <ip_address>192.0.2.18</ip_address>
	    </add>
	  </ip>
	</packet>

This packet will return an error because the nodes are not ordered as it should.
Ordering then are an easy task, just need to use the proper type:

	>>> from pleskapi import StructDict as sd
	>>> from pleskapi import odict

	>>> sd['ip']['add'] = odict()
	>>> sd['ip']['add']['ip_address'] = '192.0.2.18'
	>>> sd['ip']['add']['netmask'] = '255.255.255.0'
	>>> sd['ip']['add']['type'] = 'shared'
	>>> sd['ip']['add']['interface'] = 'eth0'
	>>> sd.dict()
	{'ip': {'add': OrderedDict([('ip_address', '192.0.2.18'), ('netmask', '255.255.255.0'), ('type', 'shared'), ('interface', 'eth0')])}}


## More examples - Basic to Complex

TODO


HISTORY.rst
.. :changelog:

History
-------

0.1.2 (2013-03-10)
++++++++++++++++++

* Fixed on PyPi again: missing MANIFEST.in
* Added PleskApiError class on package pleskapi
* Updated docs and added LICENSE file


0.1.1 (2013-03-10)
++++++++++++++++++

* Fixing on PyPi


0.1 (2013-03-10)
++++++++++++++++++

* Birth



lib/pleskapi/api.py
# -*- coding: utf-8 -*-
"""
pleskapi.api
~~~~~~~~~~~~

This module contains the main built-ins for exercing the requests to the Plesk Panel endpoint.
Note: For using OrderedDict in the requests, only with Python 2.7+ or install ordereddict from http://pypi.python.org.
"""
import requests
from converter import xml2elem, dict2xml, elem2xml, xml2dict, xml2json 
from xml.etree.ElementTree import Element as xmlobj


def build(packet, **kwargs):
    """ Build the :class:`BaseRequest <BaseRequest>`.
    :param **kwargs: Arguments that BaseRequest takes.
    """
    return BaseRequest(packet, **kwargs)

def send_packet(packet, **kwargs):
    """ Start a request to the Plesk Panel endpoint. 
    Return a :class:`BaseResponse <BaseResponse>`.
    :param **kwargs: Arguments that BaseRequest takes. 
    """
    return BaseRequest(packet, **kwargs).send()

class PleskApiError(Exception): pass

class StructDict(dict):
    def __init__(self, version):
        self.version = version

    def __getitem__(self, key):
        if key in self:
            return self.get(key)
        return self.setdefault(key, StructDict(self.version))

    def dict(self):
        packet = { 'packet' : {} }
        packet['packet'].update(self)
        packet['packet'].update({ '@version' : self.version })
        return packet

    def xml(self):
        return dict2xml(self.dict())

class BaseRequest(object):
    def __init__(self, packet, server='localhost', port='8443',
                 user='admin', password='setup', key=None, timeout=240):
        """
        Represents the data that is going to be sent to the Plesk Panel endpoint.
        :param packet: The packet for sending to the Plesk Panel endpoint. Should be only in three formats: dict, OrderedDict or xml string (without headers).
        :param server: The server IP Address or uri. Default: localhost.
        :param port: The port of the Plesk Panel. Default: 8443.
        :param user: The credentials of the Plesk Panel. Default: admin.
        :param password: The password of the user. Default: setup.
        :param key: (optional) Use a secret key instead of username and password.
        :param timeout: The timeout of the request. Default: 240.
        """
        self.server = server
        self.port = port

        if isinstance(packet, dict):
            self.packetxml = dict2xml(packet)
        elif isinstance(packet, xmlobj):
            self.packetxml = elem2xml(packet)
        elif isinstance(packet, str):
            self.packetxml = packet
        else:
            raise TypeError('Unrecognized packet type.')
        if key:
            self.headers = { 'KEY': key }
        else:
            self.headers = { 'HTTP_AUTH_LOGIN': user, 'HTTP_AUTH_PASSWD': password }
        self.headers.update({'Content-Type': 'text/xml'})
        self.timeout = timeout

    @property
    def endpoint_uri(self):
        """ The Plesk Panel endpoint api constructed. """
        return 'https://{0}:{1}/enterprise/control/agent.php'.format(self.server, self.port)

    def send(self):
        """ Start the request to the Plesk Panel endpoint. Returns xml data. 
        Return a :class:`BaseResponse <BaseResponse>`. """
        try:
            response = requests.post(self.endpoint_uri,
                                     data=self.packetxml,
                                     headers=self.headers,
                                     verify=False,
                                     timeout=self.timeout)
            if response.status_code == 500:
                raise PleskApiError('Error requesting plesk api endpoint. Check server logs for more info.')
            responsepacket = unicode(response.text.replace('\n', ''))
            return BaseResponse(responsepacket)
        #except SSLError, err: pass 
        except requests.RequestException, err: print err

class BaseResponse(object):
    def __init__(self, rpacket):
        """ Represents the response data returned from the Plesk Panel api.
        :param rpacket: The xml packet response from the Plesk Panel api.
        """
        self._rpacket = rpacket
        self._dict = self.todict()

    # TODO: Testar c/resposta retornando lista
    def response(self, bare=True):
        """ Extract the 'result' node containing only the status of the response.
        :param bare: True returns everthing which is not a dict. False returns everthing. Default: True  
        """
        self._validate()
        result = self._extract_result(self.dict, bare)
        # Return the full packet if the 'result' node is a list
        if isinstance(result, list):
            return self.dict
        version = self.dict['packet']['@version']
        return { 'result' : result, 'version' : version }

    def _validate(self):
        if 'system' in self.dict['packet']:
            raise PleskApiError(self.dict['packet']['system'])
        r = self._extract_result(self.dict)
        if not isinstance(r, list):
            if r.get('status') == 'error':
                raise PleskApiError(r)

    def _extract_result(self, packet, bare=True):
        def extract(packet):
            ext = {}
            for key in packet:
                try:
                    # Domains xml schema plesk bug.
                    if 'result' in packet['result']:
                        packet['result'] = packet['result']['result']
                    for rkey in packet['result']:
                        # If the 'result' node is a list, end loop
                        if isinstance(packet['result'], list):
                            ext = []
                            break
                        if bare:
                            if not isinstance(packet['result'][rkey], dict):
                                ext[rkey] = packet['result'][rkey]
                        else:
                            ext[rkey] = packet['result'][rkey]
                except KeyError:
                    # 'result' not find, move to the next record
                    if not isinstance(packet[key], dict):
                        continue
                    return extract(packet[key])
                except TypeError:
                    # 'ext' is not a dict, continue loop
                    continue
            return ext
        ext = extract(packet)
        if isinstance(ext, list):
            return ext
        if not ext:
            raise ValueError('Could not extract the result from the response packet: %s' % packet)
        return ext

    def element(self):
        """ The response data as xml.etree.ElementTree.Element. """
        return xml2elem(self._rpacket)

    def json(self):
        """ The response data as json. """
        return xml2json(self._rpacket)

    def todict(self):
        """ Convert to dict. """
        return xml2dict(self._rpacket)

    @property
    def dict(self):
        """ The response data as dict. """
        return self._dict


lib/pleskapi/converter.py
"""
Relies on ElementTree for the XML parsing.  This is based on
pesterfish.py but uses a different XML->JSON mapping.
The XML->JSON mapping is described at
http://www.xml.com/pub/a/2006/05/31/converting-between-xml-and-json.html

Rewritten to a command line utility by Hay Kranen < github.com/hay >

XML                              JSON
<e/>                             "e": null
<e>text</e>                      "e": "text"
<e name="value" />               "e": { "@name": "value" }
<e name="value">text</e>         "e": { "@name": "value", "#text": "text" }
<e> <a>text</a ><b>text</b> </e> "e": { "a": "text", "b": "text" }
<e> <a>text</a> <a>text</a> </e> "e": { "a": ["text", "text"] }
<e> text <a>text</a> </e>        "e": { "#text": "text", "a": "text" }

This is a mess in that it is so unpredictable -- it requires lots of testing
(e.g. to see if values are lists or strings or dictionaries).  For use
in Python this could be vastly cleaner.  Think about whether the internal
form can be more self-consistent while maintaining good external characteristics
for the JSON.
"""

import xml.etree.cElementTree as ET
from xml.etree import ElementTree
from xml.dom import minidom

try: import simplejson
except ImportError: import json as simplejson

try: from collections import OrderedDict as odict
except ImportError:
    try: import OrderedDict as odict
    except ImportError: odict = dict

def elem_to_internal(elem, strip=1):
    """Convert an Element into an internal dictionary (not JSON!). """
    d = dict()
    for key, value in elem.attrib.items():
        d['@'+key] = value

    # loop over subelements to merge them
    for subelem in elem:
        v = elem_to_internal(subelem, strip)
        tag = subelem.tag
        value = v[tag]
        try:
            # add to existing list for this tag
            d[tag].append(value)
        except AttributeError:
            # turn existing entry into a list
            d[tag] = [d[tag], value]
        except KeyError:
            d[tag] = value

    text = elem.text
    tail = elem.tail

    if strip:
        # ignore leading and trailing whitespace
        if text: text = text.strip()
        if tail: tail = tail.strip()

    if tail:
        d['#tail'] = tail

    if d:
        # use #text element if other attributes exist
        if text: d["#text"] = text
    else:
        # text is the value if no attributes
        d = text or None

    return {elem.tag: d}

def internal_to_elem(pfsh, dictype):
    """Convert an internal dictionary (not JSON!) into an Element. """
    attribs = dictype()
    text = None
    tail = None
    sublist = []
    tag = pfsh.keys()
    if len(tag) != 1:
        raise ValueError("Illegal structure with multiple tags: %s" % tag)
    tag = tag[0]
    value = pfsh[tag]
    if isinstance(value,dict):
        for k, v in value.items():
            if k[:1] == "@":
                attribs[k[1:]] = v
            elif k == "#text":
                text = v
            elif k == "#tail":
                tail = v
            elif isinstance(v, list):
                for v2 in v:
                    sublist.append(internal_to_elem({k:v2}, dictype))
            else:
                sublist.append(internal_to_elem({k:v}, dictype))
    else:
        text = value
    e = ET.Element(tag, attribs)
    for sub in sublist:
        e.append(sub)
    e.text = text
    e.tail = tail
    return e

def elem2dict(elem, strip=1):
    """Convert an ElementTree or Element into a JSON string."""
    if hasattr(elem, 'getroot'):
        elem = elem.getroot()
    return elem_to_internal(elem, strip=strip)

def xml2elem(xmlstr, strip=1):
    """Convert an XML string to an element. """
    return ET.fromstring(xmlstr)

def xml2json(xmlstring, strip=1):
    """Convert an XML string into a JSON string.
    :param xmlstring: The specified xml string to convert into a json string
    :param strip: Ignore leading and trailing whitespace. Default: 1 (ignore)
    """
    elem = ET.fromstring(xmlstring)
    return simplejson.dumps(elem2dict(elem, strip=strip))

def xml2dict(xmlstring, strip=1):
    """Convert an XML string into a dictionary.
    :param xmlstring: The specified xml string to convert into a dict
    :param strip: Ignore leading and trailing whitespace. Default: 1 (ignore)
    """
    elem = ET.fromstring(xmlstring)
    return elem2dict(elem, strip=strip)

def dict2xml(adict, prettify=False, dictype=odict):
    """Convert a dict to an xml string with headers.
    :param adict: The specified dictonary to convert into an XML
    :para prettify: Indented XML. For debbuging proporse only.
    :param dictype: The type of the dictionary which will be used to construct the xml string (dict or OrderedDict). Default: OrderedDict (odict)
    """
    elem = internal_to_elem(adict, dictype)
    elem = minidom.parseString(ElementTree.tostring(elem, encoding='UTF-8'))
    if prettify:
        return elem.toprettyxml(indent='  ', encoding='UTF-8')
    return elem.toprettyxml(indent='', encoding='UTF-8').replace('\n', '')

def elem2xml(elem):
    """Convert an element to an XML string. """
    elem = minidom.parseString(ElementTree.tostring(elem, encoding='UTF-8'))
    return elem.toprettyxml(indent='', encoding='UTF-8').replace('\n', '')

lib/pleskapi/base.py
# -*- coding: utf-8 -*-
"""
pleskapi.base
~~~~~~~~~~~~

This module contains the main built-ins for exercing the requests to the Plesk Panel endpoint.
Note: For using OrderedDict in the requests, only with Python 2.7+ or install ordereddict from http://pypi.python.org.
"""
import urllib2
from converter import xml2elem, dict2xml, elem2xml, xml2dict, xml2json, odict
from xml.etree.ElementTree import Element as xmlobj
import warnings

if dict is odict:
    warnings.simplefilter('once')
else:
    warnings.simplefilter('ignore')

def build(packet, **kwargs):
    """ Build the :class:`BaseRequest <BaseRequest>`.
    :param **kwargs: Arguments that BaseRequest takes.
    """
    return BaseRequest(packet, **kwargs)

def send_packet(packet, **kwargs):
    """ Start a request to the Plesk Panel endpoint.
    Return a :class:`BaseResponse <BaseResponse>`.
    :param **kwargs: Arguments that BaseRequest takes.
    """
    return BaseRequest(packet, **kwargs).send()

class PleskApiError(Exception): pass

class StructDict(dict):
    def __init__(self, version):
        """Helper class for building easy chain's of dict.
        dict common creation:
        >>> d = {}
        >>> d['webspace'] = {}
        >>> d['webspace']['get'] = {}
        >>> d['webspace']['get']['filter'] = {'name' : 'domain.tld'}
        >>> d
        {'webspace': {'get': {'filter': {'name': 'domain.tld'}}}}

        StructDict example:
        >>> from pleskapi import StructDict
        >>> sd = StructDict('1.6.3.5')
        >>> sd['webspace']['get']['filter'] = {'name' : 'domain.tld'}
        >>> sd
        {'webspace': {'get': {'filter': {'name': 'domain.tld'}}}}
        """
        self.version = version

    def __getitem__(self, key):
        if key in self:
            return self.get(key)
        return self.setdefault(key, StructDict(self.version))

    def dict(self):
        """Return a dict with a packet header - <packet version="version">...</packet> """
        warnings.warn("Missing OrderedDict package. Dict's are not ordered, elements SHOULD be in order or the request may fail.", ImportWarning)
        packet = { 'packet' : {} }
        packet['packet'].update(self)
        packet['packet'].update({ '@version' : self.version })
        return packet

    def xml(self, prettify=False):
        """ Return the XML representation """
        warnings.warn("Missing OrderedDict package. Dict's are not ordered, elements SHOULD be in order or the request may fail.", ImportWarning)
        return dict2xml(self.dict(), prettify)

class BaseRequest(object):
    def __init__(self, packet, server='localhost', port='8443',
                 user='admin', password='setup', key=None, timeout=240):
        """Represents the data that is going to be sent to the Plesk Panel endpoint.
        :param packet: The packet for sending to the Plesk Panel endpoint. Allowed values: dict, xml.etree.ElementTree.Element or an XML string (without headers).
        :param server: The server IP Address or uri. Default: localhost.
        :param port: The port of the Plesk Panel. Default: 8443.
        :param user: The credentials of the Plesk Panel. Default: admin.
        :param password: The password of the user. Default: setup.
        :param key: (optional) Use a secret key instead of username and password.
        :param timeout: The timeout of the request. Default: 240.
        """
        self.server = server
        self.port = port

        if isinstance(packet, dict):
            warnings.warn("Missing OrderedDict package. Dict's are not ordered, elements SHOULD be in order or the request may fail.", ImportWarning)
            self.packetxml = dict2xml(packet)
        elif isinstance(packet, type(xmlobj)):
            self.packetxml = elem2xml(packet)
        elif isinstance(packet, str):
            self.packetxml = packet
        else:
            raise TypeError('Unrecognized packet type: %s' %type(packet))
        if key:
            self.headers = { 'KEY': key }
        else:
            self.headers = { 'HTTP_AUTH_LOGIN': user, 'HTTP_AUTH_PASSWD': password }
        self.headers.update({'Content-Type': 'text/xml'})
        self.timeout = timeout

    @property
    def endpoint_uri(self):
        """The Plesk Panel endpoint API constructed. """
        return 'https://{0}:{1}/enterprise/control/agent.php'.format(self.server, self.port)

    def send(self):
        """Start the request to the Plesk Panel endpoint. Returns XML data.
        Return a :class:`BaseResponse <BaseResponse>`. """
        try:
            response = urllib2.urlopen(urllib2.Request(self.endpoint_uri, self.packetxml, self.headers), timeout=self.timeout)
            responsepacket = unicode(response.read().replace('\n', ''), 'utf-8')
            responsepacket = responsepacket.encode('utf-8')
            return BaseResponse(responsepacket)
        except urllib2.HTTPError, e:
            if e.code == 500:
                raise PleskApiError('Error requesting Plesk Api endpoint. Check server logs form more info.')
            raise

class BaseResponse(object):
    def __init__(self, rpacket):
        """Represents the response data returned from the Plesk Panel API.
        :param rpacket: The XML packet response from the Plesk Panel API.
        """
        self._rpacket = rpacket
        self._dict = self.todict()

    # TODO: Testar c/resposta retornando lista
    def response(self, bare=True):
        """Extract the 'result' node containing only the status of the response.
        :param bare: True returns everthing which is not a dict. False returns everthing. Default: True
        """
        self._validate()
        result = self._extract_result(self.dict, bare)
        # Return the full packet if the 'result' node is a list
        if isinstance(result, list):
            return self.dict
        version = self.dict['packet']['@version']
        return { 'result' : result, 'version' : version }

    def _validate(self):
        if 'system' in self.dict['packet']:
            raise PleskApiError(self.dict['packet']['system'])
        r = self._extract_result(self.dict)
        if not isinstance(r, list):
            if r.get('status') == 'error':
                raise PleskApiError(r)

    def _extract_result(self, packet, bare=True):
        def extract(packet):
            ext = {}
            for key in packet:
                try:
                    # Domains xml schema plesk bug.
                    if 'result' in packet['result']:
                        packet['result'] = packet['result']['result']
                    for rkey in packet['result']:
                        # If the 'result' node is a list, end loop
                        if isinstance(packet['result'], list):
                            ext = []
                            break
                        if bare:
                            if not isinstance(packet['result'][rkey], dict):
                                ext[rkey] = packet['result'][rkey]
                        else:
                            ext[rkey] = packet['result'][rkey]
                except KeyError:
                    # 'result' not find, move to the next record
                    if not isinstance(packet[key], dict):
                        continue
                    return extract(packet[key])
                except TypeError:
                    # 'ext' is not a dict, continue loop
                    continue
            return ext
        ext = extract(packet)
        if isinstance(ext, list):
            return ext
        if not ext:
            raise ValueError('Could not extract the result from the response packet: %s' % packet)
        return ext

    def element(self):
        """The response data as xml.etree.ElementTree.Element. """
        return xml2elem(self._rpacket)

    def json(self):
        """The response data as json. """
        return xml2json(self._rpacket)

    def todict(self):
        """Convert to dict. """
        return xml2dict(self._rpacket)

    @property
    def dict(self):
        """The response data as dict. """
        return self._dict
